src/mscorlib/src/System/Array.cs

```diff
@@ -2502,45 +2501,54 @@ sealed class SZArrayHelper {
         // This is a normal generic Enumerator for SZ arrays. It doesn't have any of the "this" stuff
         // that SZArrayHelper does.
         //
-        [Serializable] private sealed class SZGenericArrayEnumerator<T> : IEnumerator<T> {
+        [Serializable]
+        private sealed class SZGenericArrayEnumerator<T> : IEnumerator<T>
+        {
             private T[] _array;
             private int _index;
-            private int _endIndex; // cache array length, since it's a little slow.
 
-            // Passing -1 for endIndex so that MoveNext always returns false without mutating _index
-            internal static readonly SZGenericArrayEnumerator<T> Empty = new SZGenericArrayEnumerator<T>(null, -1);
+            internal static readonly SZGenericArrayEnumerator<T> Empty = new SZGenericArrayEnumerator<T>(Array.Empty<T>());
```

Comment:
```Array.Empty<T>``` incurs generic instantiation cost per T, so if you don't know that the T[] is going to be very common, in a situation like this where you're at most creating one, you're likely better off just allocating the T[0].  Worst case, you end up with at most one more T[0] allocated; best case, you save all of the generic instantiation overhead for a ```Array.Empty<T>``` you won't otherwise need.